rust/scx_utils/src/log_recorder.rs

```diff
@@ -0,0 +1,275 @@
+            let prev_value = prev_counter_values.get(key).cloned().unwrap_or(0);
+            let rate_per_second = (value - prev_value) as f64 / period_secs;
+            total_rate_per_second += rate_per_second;
+        }
+
+        info!("  {}: {} [{:.1}/s]", key_name, total, total_rate_per_second);
+
+        if key_values.len() > 1 {
+            // Sort the key_values by the counter value in descending order
+            key_values.sort_by(|a, b| b.1.cmp(&a.1));
+
+            // Log individual totals, their percentages, and rates
+            for (key, value) in key_values {
+                let prev_value = prev_counter_values.get(&key).cloned().unwrap_or(0);
+                let rate_per_second = (value - prev_value) as f64 / period_secs;
+                let percentage = (value as f64 / total as f64) * 100.0;
+                let label_value = key.labels().next().unwrap().value(); // Assuming only one label
+                info!(
+                    "    {}: {} ({:.1}%) [{:.1}/s]",
+                    label_value, value, percentage, rate_per_second
+                );
+                prev_counter_values.insert(key.clone(), value);
+            }
+        } else {
+            let (key, value) = &key_values[0];
+            prev_counter_values.insert(key.clone(), *value);
+        }
+    }
+}
+
+fn log_gauge_info(registry: &Registry<Key, AtomicStorage>) {
+    let handles = registry.get_gauge_handles();
+    let mut keys: Vec<Key> = handles.keys().cloned().collect();
+    keys.sort();
+    if keys.len() > 0 {
+        info!("Gauges:");
+    }
+    for key in keys {
+        match registry.get_gauge(&key) {
+            None => continue,
+            Some(gauge) => {
+                // Gauge values are stored as bits, so we need to convert them to f64
+                let value = f64::from_bits(gauge.load(Relaxed));
+                info!("  {}: {:.2}", key.name(), value);
+            }
+        }
+    }
+}
+
+fn log_histogram_info(registry: &Registry<Key, AtomicStorage>) {
+    let handles = registry.get_histogram_handles();
+    let mut keys: Vec<Key> = handles.keys().cloned().collect();
+
+    keys.sort();
+    if keys.len() > 0 {
+        info!("Histograms:");
+    }
+
+    for key in keys {
+        match registry.get_histogram(&key) {
+            None => continue,
+            Some(histogram) => {
+                let mut sum = 0.0;
+                let mut count = 0;
+                let mut min = 0.0;
+                let mut max = 0.0;
+                histogram.data_with(|elements| {
+                    for element in elements.iter() {
+                        sum += element;
+                        count += 1;
+
+                        if min == 0.0 || *element < min {
+                            min = element.clone();
+                        }
+
+                        if *element >= max {
+                            max = element.clone();
+                        }
+                    }
+                });
```

Comment:
Not sure if you saw it, but there's `AtomicBucket::clear_with`, which is generally preferred to `data`/`data_with` + `clear` when you're trying to atomically clear the bucket and observe all the values you cleared.

Reply:
I wasn't aware of that fn, thanks. Updated to use it.